### DISCUSSION TITLE
Easy package loading and installation

### DISCUSSION THREAD

**Discussion by aspencoyle** [RobertsLab/resources#1413] [Show and tell]
Hi all,

Here's that chunk of code I mentioned that I put at the top of all my scripts!

For scripts that don't require BiocManager, remove those lines. Otherwise, should just add a few seconds to running each script!

Also all credit to @laurahspencer for coming up with this!

```{r libraries, message = FALSE, warning=FALSE}
# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("tidyverse", "lubridate")
# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})
```

**Comment by kubu4** [comment]
Curious how this portion of code works:

```r
if(length(new.packages)) install.packages(new.packages)
```

It seems clear that it's assessing the length of the `new.packages` vector, but it doesn't seem to be comparing it to anything. So what is the point? I feel like this is a special coding "trick" that I'm not understanding. Can anyone enlighten me (I really like little coding "tricks")?

**Reply by aspencoyle** [reply]
I just tested out some equivalents, and I think with R, any number in an if() statement evaluates to TRUE, while if(0) evaluates to FALSE. 

So if length > 0, it evaluates to TRUE

Ex 1:

```
if(55){
     print("working")
}
 "working"
```

Ex 2:
```
if(0){
     print("working")
}
```

**Reply by kubu4** [reply]
Thanks for elaborating a bit.

I guess I find that line strange because if `new.packages` ends up being empty, when `install.packages(new.packages)` is run, nothing happens (i.e. no warning/error pops up). So, it seems like the `if()` statement is unnecessary.
